### PR TITLE
Get-DbaUserPermission - Exclude non-accessible databases

### PR DESCRIPTION
**Pull request by sqldeployhelmet**
If someone (me) is going to query all the servers and all the databases indiscriminately then maybe we should proactively ignore databases that are inaccessible.

## Type of Change

 - [x] Bug fix (non-breaking change, fixes # )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### Purpose

Get-DbaUserPermission throws errors when attempting to get permissions for a database that is not online (i.e. restoring etc.) This change will ensure the function ignores any databases where IsAccessible is False.

### Approach

Added a filter to the Where-Object statement when populating the $dbs array.

### CODE REVIEW

**Comment by andreasjordan** [comment]
I don't think the PR reaches the goal, as you only filter if the parameters are used.
You should also look at https://github.com/dataplat/dbatools/pull/8197/files#diff-178a92a3953c9ffd9d6513040ac43ed7a9132fc9ec384c1b5d1e9869404a2899L224-L228 - maybe just move the test up before running $endSQL and maybe not use Stop-Function but Write-Message -Level Warning.

**Comment by sqldeployhelmet** [comment]
I read the initial flow completely wrong 😅 . I went ahead and just removed all dbs where IsAccessible is not true in the initial collection and threw in a warning message if after filtering out any unrequested databases none were left. (I hope 🤞🏻.)

> I don't think the PR reaches the goal, as you only filter if the parameters are used. You should also look at https://github.com/dataplat/dbatools/pull/8197/files#diff-178a92a3953c9ffd9d6513040ac43ed7a9132fc9ec384c1b5d1e9869404a2899L224-L228 - maybe just move the test up before running $endSQL and maybe not use Stop-Function but Write-Message -Level Warning.

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:203]: With that change, we don't need the `if ($db.IsAccessible -eq $false) {` down in line 230.

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:210]: "available" or "accessible"?

**Comment by sqldeployhelmet** [comment]
@andreasjordan : the one on line 230 wasn't mine but did help me realize what was going wrong: the `$db.ExecuteNonQuery` just above (line 225 now) gets an error if the db is restoring that is un-caught...

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:229]: How can this be if you filter those dbs out in line 203?

Comment (joshs-STCU) [reply functions/Get-DbaUserPermission.ps1:229]: 🤦‍♂️ I did not mean to leave that still in there: just pulled it back out.

Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1:229]: 🤦‍♂️ I meant to pull all of that back out: just removed the last of my original "fix"

Comment (andreasjordan) [reply functions/Get-DbaUserPermission.ps1:229]: Excuse me for being so clear: You removed a good solution and introduced a nasty litte try-catch-fix just to suppress an error and not really solve the problem. There are issues with this command and we sould solve (!) them in a good way. Perhaps I represent a minority opinion, let's see what the other contributors say...

Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1:229]: I can certainly go the other direction: I just took my cues from the existing code...

Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1]: Okay, I took a stab at cleaning it up. As a very new contributor I didn't want to assume I should do more extensive re-writes. This current version moves the server level permissions to a single check (without the need for a "did we do it already flag" at the beginning of the loop on `$sqlInstances` and then moves the check to see if a db is accessible into the loop on `$dbs` and displaying a warning message if the database is not queryable...

Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1]: @andreasjordan Thanks for the positive feedback that my initial changes were on the right track. I'm just super good at second guessing myself...

**Comment by andreasjordan** [comment]
Sorry to say: You still have no idea how this thing works. Please have a look at the contents of the loaded sql file and the sql code in the begin block. Then ask yourself which code has to executed in which database - that might help to find some of the bugs in here. And if you wrap something in a try-catch with an empty catch: That means that you just don't care about the results gathered inside of the try. Is that what you want here? You don't care if results are produced or not?

**Comment by potatoqualitee** [comment]
Thank you both so much, I will merge once @andreasjordan gives the approval 🙇🏼

**Comment by sqldeployhelmet** [comment]
> Thank you both so much, I will merge once @andreasjordan gives the approval 🙇🏼

I have a couple more changes to make this morning I think that will make it better as well 😄

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:223]: Why don't you use the string "master"?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:229]: So you don't care if executing the $sql fails or not? The rest of the command would just run fine in that case? Really?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:233]: Why do you execute this query against master? Why not tempdb?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:256]: Why don't you use Stop-Function to be able to raise an exception?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:259]: Why do you have to do that?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:276]: Why do you do that (again)?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:288]: Again: Why don't you care?

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [inline functions/Get-DbaUserPermission.ps1:316]: Why don't you use Stop-Function?

**Review by andreasjordan — CHANGES_REQUESTED** [review]
And one last question: Why don't you clean up after all. You leave the stig schema behind...

I my eyes, this PR is still far away from beeing approved.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1]: I wanted to keep it consistent with how the SQL was executed through the stored function. I have dropped the $master object.. This now uses the string master and the SQL is executed via the $tempdb object.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1:228]: The code that existed in the function indicated that this try/catch was here simply becuase it _might_ fail if the objects were not cleaned up from a previous run (where a user may have control c'd their way out of the code. This simply was not part of the function I was trying to fix and it seems to have been working fine to date. If you have something specific you would like to see here please feel free to add it.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1]: Personal preference. It now executes against tempdb (see conversation further up in the code).

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1]: Because I was unaware of Stop-Function. Stop-Function is now used in the catch blocks.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1:287]: See comment above: this was working before I started making any changes and is not related to problem I am trying to solve with my update.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1]: As stated further up Stop-Function is now implemented.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1:275]: Asked and answered.

File: functions/Get-DbaUserPermission.ps1
Comment (sqldeployhelmet) [reply functions/Get-DbaUserPermission.ps1:258]: I wasn't looking to completely re-factor this function: I simply wanted to fix an issue where if run against a database that was not accessible it threw an error and failed. 

The [current version of the function](https://github.com/dataplat/dbatools/blob/development/functions/Get-DbaUserPermission.ps1) Currently acts in the same way: creating the objects, running the query and deleting the objects for each database it is running against on an instance. If there is a specific fix you would like to see here please feel free to update it or tell me specifically what you would like to see.

**Comment by sqldeployhelmet** [comment]
@andreasjordan: I've made further changes but I would like to re-iterate that the fix I am trying to implement is the ability to run this function against my estate that won't crash if it encounters a database that is not accessible. I feel like there are a number of things I am continuing to do "wrong" but that I am expected to guess at what solution would be found acceptable. I came into this as a relatively new contributor that was trying to solve a very specific issue the function was having when I tried to use it in my environment and my mindset was to change as little as possible so that it would function as expected without re-writing it. That being said:

* If there are improvements to existing code from the function that are _required_ I will happily implement those if they are explicitly defined.
* If there are changes that are desired for other reasons I am even happy to implement them if they are explicitly defined.

I would very much like to get this fixed so I (and others) can implement this to gather estate wide permissions if they like.

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [reply functions/Get-DbaUserPermission.ps1:228]: But $sql is not about cleanup, but about creating the STIG schema. If that fails, maybe an old version for another database is used and would create wrong results.

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [reply functions/Get-DbaUserPermission.ps1:258]: You are (mostly) right here. And executing the $endSQL in the contect of tempdb would clean up after the usage of STIG. But to avoid left-overs from another run, cleanup is done at the beginning, so once at the end of the instance would be enough.

File: functions/Get-DbaUserPermission.ps1
Comment (andreasjordan) [reply functions/Get-DbaUserPermission.ps1:275]: Executing $endSQL in a database context other than tempdb does: Nothing - because there is (or should be) no schema STIG.

**Comment by andreasjordan** [comment]
I have taken some time today to get a closer look at the code and did some refactorting myself. You can see at #8220 what my solution would be.

I think contributing to an open source project is not just changing code so that it does not error out on my local maschine but also about solving problems I find along the way. And understanding the general structure of the code and helper commands like Stop-Function is key to successful contributing. And personally I only change a line of code if I completly understand what the command does that I'm about to change.

**Comment by potatoqualitee** [comment]
While I agree that it's wonderful to fix issues along the way, I think that should be an ideal and not a baseline. This is for a number of reasons, including the time limitations of the contributor and reducing the barrier to contributions, especially those that help dbatools work better within various environments.

If the current code is acceptable for the scope that @sqldeployhelmet is trying to address, I'd like to merge it. 

From there, a nice-to-have PR can be submitted to address the cleanup (which I'd also appreciate for my own machines) 🙏🏼

**Comment by andreasjordan** [comment]
image

So I just used the code from this branch and it thwors a lot of warnings. Why? Because of:
```
            # now that we have server level permissions get any database level perms
            foreach ($db in $dbs | Where-Object { $_.isAccessible -eq $true }) {
                Stop-Function -Message "Database $db on $instance is not accessible." -Continue -EnableException $EnableException
            }
```

**Comment by andreasjordan** [comment]
I will stop looking at this pull request for a while.

**Comment by andreasjordan** [comment]
One more thing: There is no active test for this command. The one on AppVeyor is disabled and the one on GitHub Actions is not correct. So at the moment: This code is not tested!

**Comment by potatoqualitee** [comment]
Thank you 🙇🏼  I'll see what I can do regarding the tests. I'll also propose a change to be verbose instead of of warning

File: functions/Get-DbaUserPermission.ps1
Comment (potatoqualitee) [inline functions/Get-DbaUserPermission.ps1:268]:
```suggestion
            foreach ($db in $dbs | Where-Object isAccessible) {
```

The above is faster and prettier 💅🏼

**Comment by potatoqualitee** [comment]
Thanks again to you both. 

I just reviewed this PR as well as Andreas' PR at https://github.com/dataplat/dbatools/pull/8220 and believe that his proposed PR both addresses both of your concerns outlined in this PR, including tests. 

@sqldeployhelmet can you please take a look and let me know if his PR suits your needs? If so, I will merge that one as it's a bit more concise and does a couple more things. Once approved, I'll merge and publish to the gallery 👍🏼

**Comment by sqldeployhelmet** [comment]
@potatoqualitee : Looks good to me, assuming I am reading it correctly.

@andreasjordan : thank you for attempting to help me, I know it's not necessarily easy across so many time zones. 

Maybe if I attempt to contribute again in the future I'll pick something that is earmarked for beginners. I very much appreciate the amount of care and work that goes into dbatools!

**Pull request closed without merging**
